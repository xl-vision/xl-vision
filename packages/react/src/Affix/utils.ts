import { getBoundingClientRect, isWindow, off, on, EventMap } from '@xl-vision/utils';

export type DOMVerticalDistance = {
  top: number;
  bottom: number;
};

export const getTargetRect = (target: Window | HTMLElement): DOMVerticalDistance => {
  if (isWindow(target)) {
    return {
      top: 0,
      bottom: target.innerHeight,
    };
  }

  return getBoundingClientRect(target);
};

export function getFixedTop(
  placeholderReact: DOMVerticalDistance,
  targetRect: DOMVerticalDistance,
  offsetTop?: number,
) {
  if (offsetTop !== undefined && targetRect.top > placeholderReact.top - offsetTop) {
    return offsetTop + targetRect.top;
  }
  return undefined;
}

export function getFixedBottom(
  placeholderReact: DOMVerticalDistance,
  targetRect: DOMVerticalDistance,
  offsetBottom?: number,
) {
  if (offsetBottom !== undefined && targetRect.bottom < placeholderReact.bottom + offsetBottom) {
    const targetBottomOffset = window.innerHeight - targetRect.bottom;
    return offsetBottom + targetBottomOffset;
  }
  return undefined;
}

const TRIGGER_EVENTS = [
  'resize',
  'scroll',
  'touchstart',
  'touchmove',
  'touchend',
  'pageshow',
  'load',
] as const;

export const addTargetObserver = (target: Window | HTMLElement, fn: () => void) => {
  TRIGGER_EVENTS.forEach((name) => {
    on(target, name as keyof EventMap<Window | HTMLElement>, fn);
  });
};

export const removeTargetObserver = (target: Window | HTMLElement, fn: () => void) => {
  TRIGGER_EVENTS.forEach((name) => {
    off(target, name as keyof EventMap<Window | HTMLElement>, fn);
  });
};
