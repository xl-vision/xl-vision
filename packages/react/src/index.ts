export { styled, createGlobalStyles } from './styles';
export { default as ThemeProvider } from './ThemeProvider';
export { default as Transition } from './Transition';
export { default as CSSTransition } from './CSSTransition';
export { default as CollapseTransition } from './CollapseTransition';
export { default as TransitionGroup } from './TransitionGroup';
