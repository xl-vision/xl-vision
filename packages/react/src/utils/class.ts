export const getClasses = (element: HTMLElement) => {
  const classes = (element.className || '').trim();
  return classes.split(/\s+/);
};

export const containClass = (element: HTMLElement, className: string) => {
  return getClasses(element).some((it) => it === className);
};

export const addClass = (element: HTMLElement, className: string) => {
  if (containClass(element, className)) {
    return;
  }
  const _className = `${element.className} ${className}`;
  element.className = _className.replace(/\s+/, ' ').trim();
};

export const removeClass = (element: HTMLElement, className: string) => {
  if (!containClass(element, className)) {
    return;
  }
  const _className = element.className.replace(className, ' ');
  element.className = _className.replace(/\s+/, ' ').trim();
};
