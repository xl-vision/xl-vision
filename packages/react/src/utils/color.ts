export type RGBColor = {
  r: number;
  g: number;
  b: number;
  a?: number;
};

export type HSLColor = {
  h: number;
  s: number;
  l: number;
  a?: number;
};

export type Color = RGBColor | HSLColor;

const hexToColor: (color: string) => Color = (color) => {
  const parsedColor = color.substring(1);

  const re = new RegExp(`.{1,${parsedColor.length >= 6 ? 2 : 1}}`, 'g');
  // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
  let colors: Array<string> | RegExpMatchArray | null = parsedColor.match(re);

  if (!colors || colors.length < 3) {
    throw new Error(`The color '${color}' is illegal hex color`);
  }

  if (colors[0].length === 1) {
    colors = colors.map((n) => n + n);
  }

  const rgbColor: RGBColor = {
    r: parseInt(colors[0], 16),
    g: parseInt(colors[1], 16),
    b: parseInt(colors[2], 16),
  };

  if (colors.length > 3) {
    rgbColor.a = parseInt(colors[3], 16) / 255;
  }
  return rgbColor;
};

const rgbToColor: (color: string) => Color = (color) => {
  const matcher = color.indexOf('(');
  if (matcher === -1) {
    throw new Error(`The color '${color}' is illegal rgb color`);
  }

  const values = color.substring(matcher + 1, color.length - 1).split(',');

  if (values.length < 3) {
    throw new Error(`The color '${color}' is illegal rgb color`);
  }

  const rgbColor: RGBColor = {
    r: parseInt(values[0], 10),
    g: parseInt(values[1], 10),
    b: parseInt(values[2], 10),
  };

  if (values.length > 3) {
    rgbColor.a = parseFloat(values[3]);
  }
  return rgbColor;
};

const hslToColor: (color: string) => Color = (color) => {
  const matcher = color.indexOf('(');
  if (matcher === -1) {
    throw new Error(`The color '${color}' is illegal hsl color`);
  }

  const values = color.substring(matcher + 1, color.length - 1).split(',');

  if (values.length < 3) {
    throw new Error(`The color '${color}' is illegal hsl color`);
  }

  const hslColor: HSLColor = {
    h: parseInt(values[0], 10),
    s: parseFloat(values[1]),
    l: parseFloat(values[2]),
  };

  if (values.length > 3) {
    hslColor.a = parseFloat(values[3]);
  }
  return hslColor;
};

const hue2rgb = (p: number, q: number, t: number) => {
  // eslint-disable-next-line no-param-reassign
  if (t < 0) t += 1;
  // eslint-disable-next-line no-param-reassign
  if (t > 1) t -= 1;
  if (t < 1 / 6) return p + (q - p) * 6 * t;
  if (t < 1 / 2) return q;
  if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
  return p;
};

const hslColorToRgbColor: (color: HSLColor) => RGBColor = (color) => {
  let { h, s, l } = color;
  h /= 360;
  s /= 100;
  l /= 100;
  let r = 0;
  let g = 0;
  let b = 0;
  if (s === 0) {
    r = g = b = l; // achromatic
  } else {
    const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
    const p = 2 * l - q;
    r = hue2rgb(p, q, h + 1 / 3);
    g = hue2rgb(p, q, h);
    b = hue2rgb(p, q, h - 1 / 3);
  }

  const rgbColor: RGBColor = {
    r: Math.round(r * 255),
    g: Math.round(g * 255),
    b: Math.round(b * 255),
    a: color.a,
  };

  return rgbColor;
};

const rgbColorToHslColor: (color: RGBColor) => HSLColor = (color) => {
  let { r, g, b } = color;

  r /= 255;
  g /= 255;
  b /= 255;

  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);

  const l = (max + min) / 2;

  let h: number;
  let s: number;

  if (max === min) {
    h = s = 0;
  } else {
    const d = max - min;
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0);
        break;
      case g:
        h = (b - r) / d + 2;
        break;
      default:
        h = (r - g) / d + 4;
        break;
    }
    h /= 6;
  }

  return {
    h: h * 360,
    s: s * 100,
    l: l * 100,
    a: color.a,
  };
};

const toColor: (color: string | Color) => Color = (color) => {
  if (typeof color === 'object') {
    return color;
  }
  const parsedColor = color.trim();
  if (parsedColor.startsWith('#')) {
    return hexToColor(parsedColor);
  }
  if (parsedColor.startsWith('rgb')) {
    return rgbToColor(parsedColor);
  }
  if (parsedColor.startsWith('hsl')) {
    return hslToColor(parsedColor);
  }
  throw new Error(`The color '${color}' is not supported.`);
};

const toRgbColor: (color: string | Color) => RGBColor = (color) => {
  const obj = toColor(color);
  if ('r' in obj) {
    return obj;
  }
  return hslColorToRgbColor(obj);
};

const colorToRgb: (color: Color) => string = (color) => {
  const rgb = toRgbColor(color);
  return `rgb${rgb.a ? 'a' : ''}(${rgb.r},${rgb.g},${rgb.b}${rgb.a ? `,${rgb.a}` : ''})`;
};

const toHslColor: (color: string | Color) => HSLColor = (color) => {
  const obj = toColor(color);
  if ('h' in obj) {
    return obj;
  }
  return rgbColorToHslColor(obj);
};

const colorToHsl: (color: Color) => string = (color) => {
  const hsl = toHslColor(color);
  return `hsl${hsl.a ? 'a' : ''}(${hsl.h},${hsl.s}%,${hsl.l}%${hsl.a ? `,${hsl.a}` : ''})`;
};

const rgbNormalize = (val: number) => {
  val /= 255;
  return val <= 0.03928 ? val / 12.92 : ((val + 0.055) / 1.055) ** 2.4;
};

export const getLuminance = (color: string) => {
  let { r, g, b } = toRgbColor(color);

  r = rgbNormalize(r);
  g = rgbNormalize(g);
  b = rgbNormalize(b);

  // Truncate at 3 digits
  return Number((0.2126 * r + 0.7152 * g + 0.0722 * b).toFixed(3));
};

export const getContrastRatio = (foreground: string, background: string) => {
  const lumA = getLuminance(foreground);
  const lumB = getLuminance(background);
  return (Math.max(lumA, lumB) + 0.05) / (Math.min(lumA, lumB) + 0.05);
};

export const mix = (color1: string, color2: string, amount = 0.5) => {
  const { r: r1, g: g1, b: b1, a: a1 = 1 } = toRgbColor(color1);
  const { r: r2, g: g2, b: b2, a: a2 = 1 } = toRgbColor(color2);

  const rgbColor: RGBColor = {
    r: (r1 - r2) * amount + r2,
    g: (g1 - g2) * amount + g2,
    b: (b1 - b2) * amount + b2,
    a: (a1 - a2) * amount + a2,
  };

  return colorToRgb(rgbColor);
};

export const darken = (color: string, amount: number) => {
  const obj = toColor(color);

  if ('r' in obj) {
    return colorToRgb({
      r: obj.r * (1 - amount),
      g: obj.g * (1 - amount),
      b: obj.b * (1 - amount),
      a: obj.a,
    });
  }

  obj.l *= 1 - amount;

  return colorToHsl(obj);
};

export const lighten = (color: string, amount: number) => {
  const obj = toColor(color);

  if ('r' in obj) {
    return colorToRgb({
      r: obj.r * (1 + amount),
      g: obj.g * (1 + amount),
      b: obj.b * (1 + amount),
      a: obj.a,
    });
  }
  obj.l *= 1 + amount;

  return colorToHsl(obj);
};

export const emphasize = (color: string, coefficient = 0.1) => {
  return getLuminance(color) > 0.5 ? darken(color, coefficient) : lighten(color, coefficient);
};

export const alpha = (color: string, opacity: number) => {
  const obj = toColor(color);
  obj.a = opacity;

  if ('r' in obj) {
    return colorToRgb(obj);
  }

  return colorToHsl(obj);
};
