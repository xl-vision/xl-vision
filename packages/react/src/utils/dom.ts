// eslint-disable-next-line import/prefer-default-export
export const include = (parent: Element, child: Element) => {
  let temp: Element | null = child;
  while (temp) {
    if (temp === parent) {
      return true;
    }
    temp = temp.parentElement;
  }
  return false;
};
