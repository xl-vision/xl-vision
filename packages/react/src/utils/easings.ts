// Easing Functions - inspired from http://gizma.com/easing/

export const easeInOutCubic = (
  currentTime: number,
  totalTime: number,
  startDistance: number,
  endDistance: number,
) => {
  const t = currentTime / totalTime;

  const y = t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1;

  return y * (endDistance - startDistance) + startDistance;
};

export const linear = (
  currentTime: number,
  totalTime: number,
  startDistance: number,
  endDistance: number,
) => {
  return (currentTime / totalTime) * (endDistance - startDistance) + startDistance;
};

export const easeInCubic = (
  currentTime: number,
  totalTime: number,
  startDistance: number,
  endDistance: number,
) => {
  const t = currentTime / totalTime;

  const y = t ** 3;

  return y * (endDistance - startDistance) + startDistance;
};

export const easeOutCubic = (
  currentTime: number,
  totalTime: number,
  startDistance: number,
  endDistance: number,
) => {
  const t = currentTime / totalTime - 1;

  const y = t * t * t + 1;

  return y * (endDistance - startDistance) + startDistance;
};
