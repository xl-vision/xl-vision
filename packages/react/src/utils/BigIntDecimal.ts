import { padEnd, repeat } from '@xl-vision/utils';

export type ValueType = string | number | bigint | BigIntDecimal;

export default class BigIntDecimal {
  private origin: string;

  private _isNaN = false;

  private _integer = '';

  private _multiple = 0;

  constructor(value: ValueType) {
    if (value instanceof BigIntDecimal) {
      this.origin = value.origin;
      this._isNaN = value._isNaN;
      this._integer = value._integer;
      this._multiple = value._multiple;
      return;
    }

    let str = String(value).trim();

    this.origin = str;

    if (!str || Number.isNaN(+str)) {
      this._isNaN = true;
      return;
    }

    const isE = str.includes('e');

    if (isE) {
      const [numStr, suffix] = str.split('e');
      str = this.handleScientificNotation(numStr, +suffix);
    }

    const [integer, decimal = ''] = this.origin.split('.');

    this._multiple = decimal.length;
    this._integer = integer + decimal;
  }

  private handleScientificNotation(numStr: string, exponent: number): string {
    const isNegative = numStr.startsWith('-');
    if (isNegative) {
      numStr = numStr.slice(1);
    }
    const [integer, decimal = ''] = numStr.split('.');
    const fullNumber = integer + decimal;

    let len = exponent - decimal.length;

    let s = '';
    if (len >= 0) {
      s = fullNumber + repeat('0', len);
    } else {
      len = len + fullNumber.length;

      s =
        len > 0
          ? fullNumber.slice(0, len) + '.' + fullNumber.slice(len)
          : '0.' + repeat('0', -len) + fullNumber;
    }

    return isNegative ? '-' + s : s;
  }

  isNaN() {
    return this._isNaN;
  }

  toString() {
    if (this._isNaN) {
      return 'NaN';
    }
    if (this._multiple === 0) {
      return this._integer;
    }
    const integerPart = this._integer.slice(0, -this._multiple) || '0';
    const decimalPart = this._integer.slice(-this._multiple);
    return `${integerPart}.${decimalPart}`;
  }

  add(value: ValueType) {
    const v2 = toBigIntDecimal(value);

    if (this.isNaN() || v2.isNaN()) {
      return new BigIntDecimal(NaN);
    }

    const maxMultiple = Math.max(this._multiple, v2._multiple);

    const newBigInt =
      BigInt(padEnd(this._integer, maxMultiple - this._multiple, '0')) +
      BigInt(padEnd(v2._integer, maxMultiple - v2._multiple, '0'));

    return this.createFromBigInt(newBigInt, maxMultiple);
  }

  subtract(value: ValueType) {
    const v2 = toBigIntDecimal(value);

    if (this.isNaN() || v2.isNaN()) {
      return new BigIntDecimal(NaN);
    }

    const maxMultiple = Math.max(this._multiple, v2._multiple);

    const newBigInt =
      BigInt(padEnd(this._integer, maxMultiple - this._multiple, '0')) -
      BigInt(padEnd(v2._integer, maxMultiple - v2._multiple, '0'));

    return this.createFromBigInt(newBigInt, maxMultiple);
  }

  multiply(value: ValueType) {
    const v2 = toBigIntDecimal(value);

    if (this.isNaN() || v2.isNaN()) {
      return new BigIntDecimal(NaN);
    }

    const newMultiple = this._multiple + v2._multiple;
    const newBigInt = BigInt(this._integer) * BigInt(v2._integer);

    return this.createFromBigInt(newBigInt, newMultiple);
  }

  divide(value: ValueType, precision: number = 10) {
    const v2 = toBigIntDecimal(value);

    if (this.isNaN() || v2.isNaN() || v2._integer === '0') {
      return new BigIntDecimal(NaN);
    }

    const multiplier = BigInt('1' + repeat('0', precision));
    const dividend = BigInt(this._integer) * multiplier;
    const divisor = BigInt(v2._integer);

    const resultBigInt = dividend / divisor;
    const newMultiple = precision + this._multiple - v2._multiple;

    return this.createFromBigInt(resultBigInt, newMultiple);
  }

  toFixed(precision: number): string {
    if (this.isNaN()) {
      return 'NaN';
    }

    const integerPart = this._integer.slice(0, -this._multiple) || '0';
    let decimalPart = this._integer.slice(-this._multiple);

    decimalPart = padEnd(decimalPart, precision, '0');
    decimalPart = decimalPart.slice(0, precision);

    return precision === 0 ? integerPart : `${integerPart}.${decimalPart}`;
  }

  private createFromBigInt(value: bigint, multiple: number): BigIntDecimal {
    const str = value.toString();
    return new BigIntDecimal(this.handleScientificNotation(str, -multiple));
  }
}

const toBigIntDecimal = (value: ValueType) => {
  if (value instanceof BigIntDecimal) {
    return value;
  }
  return new BigIntDecimal(value);
};
