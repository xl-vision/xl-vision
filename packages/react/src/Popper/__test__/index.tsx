import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react';
import { triggerTransitionEnd } from 'test/utils';
import { Popper } from '@xl-vision/react';

describe('Popper', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  it('test trigger hover', () => {
    const handleOpenChange = jest.fn();

    const { container } = render(
      <Popper
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='hover'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    expect(document.body).toMatchSnapshot();

    const btn = container.querySelector('#btn')!;

    fireEvent.mouseEnter(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);

    expect(document.body).toMatchSnapshot();

    fireEvent.mouseLeave(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);

    expect(document.body).toMatchSnapshot();
  });
  it('test trigger click', async () => {
    const handleOpenChange = jest.fn();

    const { container } = render(
      <Popper
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='click'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    expect(document.body).toMatchSnapshot();

    const btn = container.querySelector('#btn')!;

    const user = userEvent.setup({ delay: null });

    await user.click(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);
    expect(document.body).toMatchSnapshot();

    document.body.click();

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);
    expect(document.body).toMatchSnapshot();
  });
  it('test trigger contextMenu', () => {
    const handleOpenChange = jest.fn();

    const { container } = render(
      <Popper
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='contextMenu'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    expect(document.body).toMatchSnapshot();

    const btn = container.querySelector('#btn')!;

    fireEvent.contextMenu(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);
    expect(document.body).toMatchSnapshot();

    document.body.click();

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);
    expect(document.body).toMatchSnapshot();
  });
  it('test trigger focus', () => {
    const handleOpenChange = jest.fn();

    const { container } = render(
      <Popper
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='focus'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    expect(document.body).toMatchSnapshot();

    const btn = container.querySelector('#btn')!;

    fireEvent.focus(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);
    expect(document.body).toMatchSnapshot();

    fireEvent.blur(btn);

    act(() => {
      jest.runAllTimers();
    });

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);
    expect(document.body).toMatchSnapshot();
  });
  it('test trigger custom', async () => {
    // 外部修改不触发onOpenChange
    const handleOpenChange = jest.fn();

    const { container, rerender } = render(
      <Popper
        open={false}
        popup={<div id='popup'>popup</div>}
        popupContainer={() => document.body}
        trigger={false}
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );
    expect(handleOpenChange.mock.calls.length).toBe(0);
    expect(document.querySelector('#popup')).toBeNull();

    const btn = container.querySelector('#btn')!;

    fireEvent.mouseEnter(btn);

    await triggerTransitionEnd();

    expect(handleOpenChange.mock.calls.length).toBe(0);
    expect(document.querySelector('#popup')).toBeNull();

    rerender(
      <Popper
        open={true}
        popup={<div id='popup'>popup</div>}
        popupContainer={() => document.body}
        trigger={false}
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    await triggerTransitionEnd();

    expect(handleOpenChange.mock.calls.length).toBe(0);
    let el = document.querySelector('#popup')!.parentElement!;
    expect(el).not.toBeNull();
    expect(el.style.display).toBe('');

    rerender(
      <Popper
        open={false}
        popup={<div id='popup'>popup</div>}
        popupContainer={() => document.body}
        trigger={false}
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    await triggerTransitionEnd();

    expect(handleOpenChange.mock.calls.length).toBe(0);

    el = document.querySelector('#popup')!.parentElement!;

    expect(el).not.toBeNull();

    expect(el.style.display).toBe('none');
  });
  it('test disablePopupEnter', async () => {
    const handleOpenChange = jest.fn();
    const { container, rerender } = render(
      <Popper
        className='popup'
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='hover'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    const btn = container.querySelector('#btn')!;

    fireEvent.mouseEnter(btn);

    await triggerTransitionEnd();

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);

    fireEvent.mouseLeave(btn);

    const popup = document.querySelector('div.popup')!;

    fireEvent.mouseEnter(popup);

    await triggerTransitionEnd();

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);

    fireEvent.mouseLeave(popup);

    await triggerTransitionEnd();

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);

    rerender(
      <Popper
        className='popup'
        hoverOptions={{ disablePopperEnter: true }}
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='hover'
        onOpenChange={handleOpenChange}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    fireEvent.mouseEnter(btn);

    await triggerTransitionEnd();

    expect(handleOpenChange).toHaveBeenLastCalledWith(true);

    fireEvent.mouseLeave(btn);
    fireEvent.mouseEnter(popup);

    await triggerTransitionEnd();

    expect(handleOpenChange).toHaveBeenLastCalledWith(false);
  });

  it('test mountOnShow', () => {
    const { container } = render(
      <Popper
        id='popup'
        mountOnShow={false}
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='hover'
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    act(() => {
      jest.runAllTimers();
    });

    let el = document.querySelector('#popup');

    expect(el).not.toBe(null);

    const btn = container.querySelector('#btn')!;

    fireEvent.mouseEnter(btn);

    act(() => {
      jest.runAllTimers();
    });

    el = document.querySelector('#popup');
    expect(el).not.toBe(null);
  });

  it('test unmountOnHide', async () => {
    const { container } = render(
      <Popper
        id='popup'
        popup={<div>popup</div>}
        popupContainer={() => document.body}
        trigger='hover'
        unmountOnHide={true}
      >
        <button id='btn'>button</button>
      </Popper>,
    );

    await triggerTransitionEnd();

    let el = document.querySelector('#popup');

    expect(el).toBe(null);

    const btn = container.querySelector('#btn')!;

    fireEvent.mouseEnter(btn);

    await triggerTransitionEnd();

    el = document.querySelector('#popup');
    expect(el).not.toBe(null);

    fireEvent.mouseLeave(btn);

    await triggerTransitionEnd();
    jest.runAllTimers();

    el = document.querySelector('#popup');
    expect(el).toBe(null);
  });
});
