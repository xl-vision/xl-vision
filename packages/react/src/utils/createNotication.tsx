import {
  NoticationContainerType,
  NoticationHookReturnType,
  NoticationOptions,
  NoticationProps,
  useNotication,
} from '@xl-vision/hooks';

import { isProduction, noop } from '@xl-vision/utils';
import { ComponentType, createRef, forwardRef, useImperativeHandle, useState } from 'react';
import { Root, createRoot } from 'react-dom/client';

type MethodNoticationRef<P, NCP> = {
  instance: ReturnType<typeof useNotication<P, NCP>>[0];
  sync: () => void;
};

const createNotication = <P, NCP>(
  Notication: ComponentType<NoticationProps<P>>,
  NoticationContainer: NoticationContainerType<NCP>,
  defaultGlobalConfig: NoticationOptions<NCP>,
) => {
  const globalConfigRef = {
    value: {
      ...defaultGlobalConfig,
    },
  };

  const MethodRefNotication = forwardRef<MethodNoticationRef<P, NCP>>((_, ref) => {
    const [noticationListState, setNoticationListState] = useState(globalConfigRef.value);

    const [methods, holder] = useNotication(Notication, NoticationContainer, noticationListState);

    useImperativeHandle(ref, () => {
      return {
        instance: methods,
        sync() {
          setNoticationListState(globalConfigRef.value);
        },
      };
    });

    return holder;
  });

  if (!isProduction) {
    MethodRefNotication.displayName = 'MethodRefNotication';
  }

  const noticationRef = createRef<MethodNoticationRef<P, NCP>>();

  let rootEl: HTMLElement | undefined;
  let root: Root | undefined;
  let count = 0;

  const destroyDOM = () => {
    if (root) {
      root.unmount();
      root = undefined;
    }
    if (rootEl) {
      if (rootEl.parentNode) {
        rootEl.parentNode.removeChild(rootEl);
      }
      rootEl = undefined;
    }
  };

  const open = (props: NoticationProps<P>): NoticationHookReturnType<P> => {
    const currentProps = {
      ...props,
    };

    let hookMethods: NoticationHookReturnType<P> | undefined;

    let promiseResolve: () => void | undefined;

    if (!rootEl) {
      const div = document.createElement('div');
      document.body.appendChild(div);
      rootEl = div;

      root = createRoot(div);
      root.render(<MethodRefNotication ref={noticationRef} />);
    }

    count++;

    Promise.resolve()
      .then(() => {
        hookMethods = noticationRef.current?.instance.open({
          ...currentProps,
          onAfterClosed() {
            count--;
            if (count <= 0) {
              destroyDOM();
            }
            currentProps.onAfterClosed?.();
            promiseResolve?.();
          },
        });
      })
      .catch(noop);

    const promise = new Promise<void>((resolve) => {
      promiseResolve = resolve;
    }) as NoticationHookReturnType<P>;

    promise.update = (updateProps) => hookMethods?.update(updateProps);
    promise.destroy = () => hookMethods?.destroy();
    promise.isDestroyed = () => hookMethods?.isDestroyed() || false;

    return promise;
  };

  const setGlobalConfig = (config: Partial<NoticationOptions<NCP>>) => {
    globalConfigRef.value = {
      ...globalConfigRef.value,
      ...config,
    };

    noticationRef.current?.sync();
  };

  return {
    setGlobalConfig,
    open,
    destroyAll: () => noticationRef.current?.instance.destroyAll(),
  };
};

export default createNotication;
