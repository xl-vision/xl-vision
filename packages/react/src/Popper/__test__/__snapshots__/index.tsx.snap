// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper test trigger click 1`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger click 2`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger click 3`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger contextMenu 1`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger contextMenu 2`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger contextMenu 3`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger focus 1`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger focus 2`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger focus 3`] = `
Array [
  <button
    id="btn"
  >
    button
  </button>,
  <div
    aria-hidden="true"
    class="xl-popper"
    style="position: absolute; display: none;"
  >
    <div
      class="xl-popper__inner"
      style="position: relative; display: none;"
    >
      <div>
        popup
      </div>
    </div>
  </div>,
]
`;

exports[`Popper test trigger hover 1`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger hover 2`] = `
<button
  id="btn"
>
  button
</button>
`;

exports[`Popper test trigger hover 3`] = `
Array [
  <button
    id="btn"
  >
    button
  </button>,
  <div
    aria-hidden="true"
    class="xl-popper"
    style="position: absolute; display: none;"
  >
    <div
      class="xl-popper__inner"
      style="position: relative; display: none;"
    >
      <div>
        popup
      </div>
    </div>
  </div>,
]
`;
